context/config_file.go

```diff
@@ -10,57 +10,161 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
-const defaultHostname = "github.com"
-
-type configEntry struct {
-	User  string
-	Token string `yaml:"oauth_token"`
-}
-
-func parseOrSetupConfigFile(fn string) (*configEntry, error) {
-	entry, err := parseConfigFile(fn)
+func parseOrSetupConfigFile(fn string) (Config, error) {
+	config, err := ParseConfig(fn)
 	if err != nil && errors.Is(err, os.ErrNotExist) {
 		return setupConfigFile(fn)
 	}
-	return entry, err
+	return config, err
+}
+
+func ParseDefaultConfig() (Config, error) {
+	return ParseConfig(configFile())
 }
 
-func parseConfigFile(fn string) (*configEntry, error) {
+var ReadConfigFile = func(fn string) ([]byte, error) {
 	f, err := os.Open(fn)
 	if err != nil {
 		return nil, err
 	}
 	defer f.Close()
-	return parseConfig(f)
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
 }
 
-// ParseDefaultConfig reads the configuration file
-func ParseDefaultConfig() (*configEntry, error) {
-	return parseConfigFile(configFile())
+var WriteConfigFile = func(fn string, data []byte) error {
+	cfgFile, err := os.OpenFile(configFile(), os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600) // cargo coded from setup
+	if err != nil {
+		return err
+	}
+	defer cfgFile.Close()
+
+	n, err := cfgFile.Write(data)
+	if err == nil && n < len(data) {
+		err = io.ErrShortWrite
+	}
+
+	return err
 }
 
-func parseConfig(r io.Reader) (*configEntry, error) {
-	data, err := ioutil.ReadAll(r)
+var BackupConfigFile = func(fn string) error {
+	return os.Rename(fn, fn+".bak")
+}
+
+func parseConfigFile(fn string) ([]byte, *yaml.Node, error) {
+	data, err := ReadConfigFile(fn)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	var config yaml.Node
-	err = yaml.Unmarshal(data, &config)
+
+	var root yaml.Node
+	err = yaml.Unmarshal(data, &root)
 	if err != nil {
-		return nil, err
+		return data, nil, err
 	}
-	if len(config.Content) < 1 {
-		return nil, fmt.Errorf("malformed config")
+	if len(root.Content) < 1 {
+		return data, &root, fmt.Errorf("malformed config")
 	}
-	for i := 0; i < len(config.Content[0].Content)-1; i = i + 2 {
-		if config.Content[0].Content[i].Value == defaultHostname {
-			var entries []configEntry
-			err = config.Content[0].Content[i+1].Decode(&entries)
-			if err != nil {
-				return nil, err
+	if root.Content[0].Kind != yaml.MappingNode {
+		return data, &root, fmt.Errorf("expected a top level map")
+	}
+
+	return data, &root, nil
+}
+
+func isLegacy(root *yaml.Node) bool {
+	for _, v := range root.Content[0].Content {
+		if v.Value == "hosts" {
```

Comment:
Could we instead detect the legacy format by checking whether there is a top-level key called `github.com`?

The rationale is that checking for a `hosts` key might not be forward-compatible. If we extract OAuth credentials to external credential store, then the `hosts` section can disappear altogether from the config file.